cpp/src/strings/capitalize.cu

```diff
@@ -0,0 +1,343 @@
+    }
+
+    __host__ __device__
+    int32_t const* get_offsets(void) const
+    {
+      return d_offsets_;
+    }
+
+    __host__ __device__
+    char* get_chars(void)
+    {
+      return d_chars_;
+    }
+
+    __device__
+    char_info get_char_info(char_utf8 chr) const
+    {
+      uint32_t code_point = detail::utf8_to_codepoint(chr);
+      detail::character_flags_table_type flag = code_point <= 0x00FFFF ? d_flags_[code_point] : 0;
+      return char_info{code_point,flag};
+    }
+
+    __device__
+    char_utf8 convert_char(char_info const& info) const
+    {
+      return detail::codepoint_to_utf8(d_case_table_[info.first]);
+    }
+  private:
+    column_device_view const d_column_;
+    character_flags_table_type const* d_flags_;
+    character_cases_table_type const* d_case_table_;
+    int32_t const* d_offsets_;
+    char* d_chars_;
+  };
+
+
+  //functor for probing string capitalization
+  //requirements:
+  //(private inheritance to prevent polymorphic use,
+  // a requirement that came up in code review)
+  //
+  struct probe_capitalize: private probe_execute_base
+  {
+    probe_capitalize(column_device_view const d_column,
+                     character_flags_table_type const* d_flags,
+                     character_cases_table_type const* d_case_table):
+      probe_execute_base(d_column, d_flags, d_case_table)
+    {  
+    }
+    
+     __device__
+     int32_t operator()(size_type idx) const {
+       if( get_column().is_null(idx) )
+         return 0; // null string
+      
+       string_view d_str = get_column().template element<string_view>(idx);
+       int32_t bytes = 0;
+      
+       for( auto itr = d_str.begin(); itr != d_str.end(); ++itr ) {
+         auto the_chr = *itr;
+
+         auto pair_char_info = get_char_info(the_chr);
+         detail::character_flags_table_type flag = pair_char_info.second;
+         
+         if( (itr == d_str.begin()) ? IS_LOWER(flag) : IS_UPPER(flag) )
+           the_chr = convert_char(pair_char_info);
+         bytes += detail::bytes_in_char_utf8(the_chr);
+       }
+       return bytes;
+    }
+  };
+
+  //functor for executing string capitalization:
+  //(private inheritance to prevent polymorphic use,
+  // a requirement that came up in code review)
+  //
+  struct execute_capitalize: private probe_execute_base
+  {
+    execute_capitalize(column_device_view const d_column,
+                       character_flags_table_type const* d_flags,
+                       character_cases_table_type const* d_case_table,
+                       int32_t const* d_offsets,
+                       char* d_chars):
+      probe_execute_base(d_column, d_flags, d_case_table, d_offsets, d_chars)
+    {
+    }
+    
+    __device__
+    int32_t operator()(size_type idx) {
+      if( get_column().is_null(idx) )
+        return 0; // null string
+      
+      string_view d_str = get_column().template element<string_view>(idx);
+      char* d_buffer = get_chars() + get_offsets()[idx];
```

Comment:
The `get_chars()` and `get_offsets()` accessors are only needed here so perhaps they could be replaced with a base class member function that simply returns the output buffer pointer. Something like:
```
    __device__ char* get_output_ptr(size_type idx)
    {
        return d_chars && d_offsets ? d_chars + d_offsets[idx] : nullptr;
    }
```
Then `get_chars()` and `get_offsets()` would no longer be needed.

Reply:
Done.